### ISSUE TITLE
Open all quests at same point

### ISSUE THREAD

**Issue by wolfgang42** [streetcomplete/StreetComplete#2710]
**Use case**
There are often several quests attached to the same point (for example, a bus stop that wants a number, whether there are benches, whether there is a shelter, etc). Answering all of the quests requires extra taps on the quest marker.

**Proposed Solution**

After completing a quest, look for another quest at the same place and automatically open it if one exists.

**Related discussion**
This is related to #124, but hopefully simpler to implement as it doesn't require any UI design.

#164 apparently implemented something like this, but it was removed again (https://github.com/streetcomplete/StreetComplete/issues/1969#issuecomment-723457563; see also #2360) and I haven't been able to track down any discussion as to why.

**Comment by smichel17** [comment]
> I haven't been able to track down any discussion as to why.

Nor I; it would be useful to know for considering future designs.

**Comment by westnordost** [comment]
I removed it again in some version because it wasn't working that well. It was an unwanted automatism.
You could download that old version and try out yourself (don't remember which version it was though), but in my experience, it just did not work well.

**Comment by wolfgang42** [comment]
I see now it was removed in 25dcafd6c7c9c985dca06d7a6e81519d0a759621, but the comment there (“it does not feel good”) does not seem any more illuminating as to what the trouble was. I did try downgrading to v6.1, but unfortunately that version seems to be unable to download the map tiles any more which makes it impossible to test properly.

If this is going to be tagged as ‘wontfix’, can you at least expound on what the problems were with the discarded implementation? It's very difficult to offer suggestions for improvements when the only explanation offered is “we made an attempt but it was bad.”

**Comment by matkoniecz** [comment]
Well, note in this case it is @westnordost project and if he removed already implemented and fully functional feature then it is extremely unlikely to get back just because someone made some suggestion.

If someone really wants it they can revert that removal commit (not sure how complicated it will be) or patch old version with new tiles (no idea how hard it is) to experiment. Though note that even if you discover that you personally like it then it is likely not reappearing anyway.

My taste/preference decisions are not always matching, but 

- overall esthetic/feel of StreetComplete is magnificent, one of better among open source tools
-  @westnordost is a critical resource in StreetComplete
- differences are basically always about minor things
- in noticeable part of cases where I initially clearly disagreed (1/5?) I get convinced anyway after some testing

Personally I liked this feature but it remember that it was sometimes clunky, confusing and I have not liked it enough to keep it in my personal fork.

In short: “it does not feel good” is a perfectly valid explanation in such case.

**Comment by westnordost** [comment]
I can not offer any more than that. It felt jagged and like an unwanted automatism. You are free to try implementing it and trying for yourself.

On April 2, 2021 5:36:25 AM GMT+02:00, Wolfgang Faust ***@***.***> wrote:
>I see now it was removed in 25dcafd6c7c9c985dca06d7a6e81519d0a759621,
>but the comment there (“it does not feel good”) does not seem any more
>illuminating as to what the trouble was. I did try downgrading to v6.1,
>but unfortunately that version seems to be unable to download the map
>tiles any more which makes it impossible to test properly.>
>>
>If this is going to be tagged as ‘wontfix’, can you at least expound on
>what the problems were with the discarded implementation? It's very
>difficult to offer suggestions for improvements when the only
>explanation offered is “we made an attempt but it was bad.”>
>>
>-- >
>You are receiving this because you modified the open/close state.>
>Reply to this email directly or view it on GitHub:>
>https://github.com/streetcomplete/StreetComplete/issues/2710#issuecomment-812299095

-- 
Diese Nachricht wurde von meinem Android-Gerät mit K-9 Mail gesendet.

**Comment by westnordost** [comment]
...reading it again: Maybe it feels better now, if one f.e. showed a different transition animation (instead of close -> open) and since all quests are now available without necessity to re-download again, which makes it predictable which quest(s) will appear next.

So, please, try it out and of you like it,  create a PR to show it to others. If people like it, it can be included again.

**Comment by westnordost** [comment]
> How to best describe it: Like [some carousel here](https://m3.material.io/components/carousel/overview), but with "cards" spawning most of the width, as you usually want to see the current quest and only a small part may indicate the next or previous one.

Hm, that sounds kind of cool. Maybe it could be a landscape-mode / tablet / big phone screen only feature. For small phones, 360px in width is sometimes barely enough to fit the quest UI.
I will keep this suggestion in mind when I remake the UI of these. (It would come at the very very end of the whole effort.)

**Comment by Amaanprobably** [comment]
> Yes, I would not suggest such an animation at all, this is confusing. Awesome would be some sliding animation or even a slider in general allowing you to swipe right or left for the next/previous quest. 

A swipe right or left for next/previous quest was what I had in mind too before coming across this issue (or due to discoverability reasons, `left/right` arrow buttons perhaps?). 

> How to best describe it: Like [some carousel here](https://m3.material.io/components/carousel/overview), but with "cards" spawning most of the width, as you usually want to see the current quest and only a small part may indicate the next or previous one.

Thinking about it more, I don't think `portrait mode` and `landscape mode` would necessarily need different implementations. 

We could have a `single carousel/viewport` that is wider than the screen (roughly 2.5–3× the normal quest width), with the current quest always centered at its usual 1.0× size and the adjacent quests at 0.6×. I'd also leave around 0.2× of spacing on each side of the adjacent quests, so they don't look cramped next to the current quest.

In `portrait mode`, the `viewport` would effectively only show the centered quest, so swiping would bring the adjacent quest into view and expand it to the center. In `landscape mode`, the adjacent quests would be fully visible and clickable, allowing the user to either swipe or tap the previous/next quest to move it into the center.

That also means we could keep the same carousel geometry and transition in both orientations; the only real difference would be how much of it is visible. Also I don't think we would need to create new `cards` for it, as the current quest forms already fulfill all requirements.

Obviously, swiping/clicking should only be enabled when there actually is a valid previous/next quest to navigate to. The `Quest Form` itself would still determine whether anything about the quest `hasChanges` after answering it before allowing navigation.

> The next quest would automatically be swiped too, once the previous one is solved.

After completing a quest, it would be better for it to `fade into the center` as that would retain consistency in animation across both **Landscape** and **Portrait** mode (with swiping it wouldn't have a place to disappear to in **Landscape** mode) & having the `next` become the `current`.

> I will keep this suggestion in mind when I remake the UI of these but I am not willing to reopen this issue as it is kind of unicorny right now. (It would come at the very very end of the whole effort.)

@westnordost Would you be open to re-opening it now? I would love to have it assigned to me, unless you'd like to take it on yourself. 
Obviously once we've agreed on the exact behavior and what's in/out of scope.
> Also currently the quest form is `leftAligned` in `Landscape mode`, which would change with the suggested approach.

**Comment by westnordost** [comment]
Hmm. Regarding landscape mode, if I understand you right, you'd fill that space with the carousel. Maybe it looks like there was space for it, but there isn't. The space is necessary to show the quest geometry etc.. This is the reason why quest forms are not full screen but have a certain peek height.
Also, it would be weird to see several quest forms at the same time. Not even sure if this could cause unintended side-effects.

As written before, you are free to try out what it feels like if the next quest just slides (or fades) in from the side when one quest is solved. I am not sure this behavior would be positive even with non-janky animation, so no guarantee for merging it.

As for offering a UI to switch between quests, I think a swipe left/right is the wrong approach. First of all, as you noticed, when changes have been made, it'll not feel good to be detered in such a way from swiping. Second, why would one want to see the whole quest **form** before swiping on to the next quest? I think the icon would be enough. Third, it is inconvenient to reach the quest one is looking to solve. You may have to swipe multiple times until you find the form you are looking for.

An easy interface could be to open a popup or circle menu that shows all available quests. The easiest method would be to do that on a long press. More discoverable would be that if the same element has several quests, a number on/next to the quest pin is displayed after the form has opened, and when clicking on it again, that menu opens. SCEE does also do something similar to this, I think, just listing all icons of other quests for this element directly above the quest form or something.

**Comment by Amaanprobably** [comment]
> Hmm. Regarding landscape mode, if I understand you right, you'd fill that space with the carousel. Maybe it looks like there was space for it, but there isn't. The space is necessary to show the quest geometry etc.. This is the reason why quest forms are not full screen but have a certain peek height. Also, it would be weird to see several quest forms at the same time. Not even sure if this could cause unintended side-effects.

Ahh, you're right. It'd indeed not fit well, nor look good now that I imagine it.

> As for offering a UI to switch between quests, I think a swipe left/right is the wrong approach. First of all, as you noticed, when changes have been made, it'll not feel good to be detered in such a way from swiping....

Hmm.. true.

> An easy interface could be to open a popup or circle menu that shows all available quests. The easiest method would be to do that on a long press. More discoverable would be that if the same element has several quests, a number on/next to the quest pin is displayed after the form has opened, and when clicking on it again, that menu opens.

So if I understand quickly, the flow is: Click on quest icon -> form opens, icon shows a number next to it -> Click on icon again -> Popup opens -> Select a quest (*from a scrollable list ig?*) -> The quest form changes to that of selected quest.

That does sound good in practice, but what about the behavior after completion of the **selected quest**? 
Would the quest after the **completed quest** be the one that is displayed, according to its natural ordering? (rather than the one after the **initially displayed quest**)

> SCEE does also do something similar to this, I think, just listing all icons of other quests for this element directly above the quest form or something.

I'd take a look into it and come back with a more polished idea in mind.

**Comment by westnordost** [comment]
> So if I understand quickly, the flow is: Click on quest icon -> form opens, icon shows a number next to it -> Click on icon again -> Popup opens -> Select a quest (from a scrollable list ig?) -> The quest form changes to that of selected quest.

Yes, the number would be shown on/next to the quest pin on the map. I'd slightly prefer a circle menu over a dropdown. Note that it is a rather rare use case that you actually need to switch between quests, so any UI needs to be **unobstrusive** or maybe even invisible while not used (e.g. hide behind long-press).

> That does sound good in practice, but what about the behavior after completion of the selected quest?
>  Would the quest after the completed quest be the one that is displayed, according to its natural ordering? (rather than the one after the initially displayed quest)

I see this as two separate features:
1. the ability to switch between open quests on the same element, covering the rare use case that one quest might not be answerable but another one is
2. the behavior to automatically open the next quest on the same element when the first has been solved. Note that since some quests are dependent on another (one quest is only created after another has been solved), e.g. building type → housenumber, the second quest would not actually be shown in feature # 1 before the form is opened automatically.
  Given that I briefly had the behavior of auto-opening before, no guarantees in merging it when implemented. The issue was *not only* that the animation was janky, but also that it felt like an unwanted automatism. Often, I didn't *want* to answer the next quest right away.
  E.g. housenumbers are only visible from more or less in front of the house. Building types are already visible way ahead. So, when going down a road, where both building types and housenumbers are missing, one can answer the first quest for dozens of buildings already. In that case, if the housenumber quest pops up after every answer, that would be very vexing.

**Comment by Amaanprobably** [comment]
Hmm.... Well, in any case, I'd have to take a look into optimizing the quest loading times first, otherwise the animation would always stay *janky* regardless of how I implement this feature.

**Comment by rugk** [comment]
> Also, it would be weird to see several quest forms at the same time. Not even sure if this could cause unintended side-effects.

IMHO, one could grey out the "secondary" (= other) quests or somehow disable them, while the primary one is shown as is.

> the behavior to automatically open the next quest on the same element when the first has been solved

I agree here, that does not sound good. After all, when a quest appears you also have an icon on the map you can _actively click on_ which IMHO is no problem for a new quest. After all, it does not hide a different quest one cannot solve already.
Think of computer game quests, it's the same mechanism there. No quest auto-opens.

>  I'd slightly prefer a circle menu over a dropdown

I like the idea. Initially I think I'd also tend to a circle menu, which could look neat (with multiple quest icons), maybe also ordered like a flower bouquet even (I don't know if that is a UI paradigm, I just called it like this). But one has to see that in practice, also to see how good it actually works.
